### PR TITLE
Support lazy loaded modules in NgModule parsing

### LINKED ISSUE #916: Support lazy loaded modules in NgModule
When a lazy loaded module load into the application, it does not appear in the NgModule tab.

We should refresh the list when a module is loaded.

### PR DESCRIPTION
**Pull request by stevenkampen**
resolves #916

### CODE REVIEW

File: src/backend/backend.ts [hunk 1]
```diff
@@ -67,17 +72,33 @@ const messageBuffer = new MessageQueue<Message<any>>();
 
 /// NOTE(cbond): We collect roots from all applications (mulit-app support)
 let previousTree: MutableTree,
+  previousRoutes: Array<Route>,
   previousCount: number,
-  initialModules: { [key: string]: any },
   onMouseOver,
   onMouseDown;
 
+const parsedModulesData: NgModulesRegistry = {
```
Comment (stevenkampen) [inline src/backend/backend.ts]: NgModules data object is created initially and added to with each parsing.

File: src/backend/backend.ts [hunk 2]
```diff
@@ -67,17 +72,33 @@ const messageBuffer = new MessageQueue<Message<any>>();
...
+  modules: {},
+  names: [],
+  configs: {},
+  tokenIdMap: {},
+};
+
+const sendNgModulesMessage = () => {
```
Comment (stevenkampen) [inline src/backend/backend.ts]: Encapsulated logic to send current NgModule data to the frontend

File: src/backend/backend.ts [hunk 3]
```diff
@@ -106,8 +127,34 @@ const updateComponentTree = (roots: Array<any>) => {
   previousCount = count;
 };
 
-const updateRouterTree = (routes: Array<MainRoute>) => {
-  messageBuffer.enqueue(MessageFactory.routerTree(routes));
+const updateLazyLoadedNgModules = (routers) => {
```
Comment (stevenkampen) [inline src/backend/backend.ts]: Called on router change, this function parses any newly loaded NgModules from the routers and updates the frontend.

File: src/backend/backend.ts [hunk 4]
```diff
@@ -106,8 +127,34 @@ const updateComponentTree = (roots: Array<any>) => {
...
+};
+
+const updateRouterTree = () => {
+  const routers: Array<any> = routerTree();
+  const parsedRoutes = routers.map(parseRoutes);
+
+  let routesChanged = !previousRoutes ? true : false;
+
+  if (previousRoutes) {
+    const changes = compare(previousRoutes, parsedRoutes);
```
Comment (stevenkampen) [inline src/backend/backend.ts]: Use `compare` to check for changes in the router and only send data to the frontend if there are any.

File: src/backend/backend.ts [hunk 5]
```diff
@@ -227,9 +274,6 @@ const messageHandler = (message: Message<any>) => {
         message.content.path,
         message.content.value));
 
-    case MessageType.RouterTree:
```
Comment (stevenkampen) [inline src/backend/backend.ts]: This was never required. Here, the backend responds to messages from the frontend, so this doesn't make sense.

File: src/backend/utils/parse-modules.ts [hunk 6]
```diff
@@ -1,10 +1,18 @@
 import {classDecorators, componentMetadata} from '../../tree/decorators';
 import {functionName} from '../../utils/function-name';
+import {Route} from '../utils/parse-router';
 
 export const AUGURY_TOKEN_ID_METADATA_KEY = '__augury_token_id';
 
 declare const ng;
 
+export interface NgModulesRegistry {
```
Comment (stevenkampen) [inline src/backend/utils/parse-modules.ts]: Added interface to describe parsed router data.

File: src/communication/message-factory.ts [hunk 7]
```diff
@@ -6,7 +6,7 @@ import {
 } from './message';
 
 import {
-  MainRoute,
+  Route,
```
Comment (stevenkampen) [inline src/communication/message-factory.ts]: Removed `MainRoute` and replaced usage with `Route`.

**Comment by stevenkampen** [comment]
@rajinder-yadav @tyllrb If you could make a start at reviewing this logic, that would be great. I may still tweak one or two things.

**Comment by stevenkampen** [comment]
@rajinder-yadav I added a small improvement here in another commit. Can you give it a quick review and approve if it's ok?

**Comment by rajinder-yadav** [comment]
:ship:

**Pull request merged**